---
title: 'Lab 2: Enrichment Analyses'
subtitle: "BMI 206"
author: "Wanjun Gu"
date: "10/27/2023"
output: html_document
---
<br>
<br>

Import needed packages and set a random number seed
```{r, message=FALSE}
#load packages
suppressPackageStartupMessages({
  suppressWarnings({
    
    
    if(!require("BiocManager", quietly = TRUE)){
      install.packages("BiocManager")
    }
    
    if(!require("regioneR")){
      BiocManager::install("regioneR")
      library("regioneR")
    }
    
    if(!require("BSgenome.Hsapiens.UCSC.hg19")){
      BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
      library("BSgenome.Hsapiens.UCSC.hg19")
    }
    
    if(!require("dplyr")){
      BiocManager::install("dplyr")
      library("dplyr")
    }
  })
})

set.seed(492357816)
```
<br>

Read in BED formatted region files: all tested regions and two sets of positives
Note: these are in hg19 human genome assembly coordinates
```{r,eval=TRUE}
all=toGRanges(read.table("all.bed",sep="\t"))
hits1=toGRanges(read.table("hits1.bed",sep="\t"))
hits2=toGRanges(read.table("hits2.bed",sep="\t"))
```
<br>

__Q1. How many regions are in hits1? How many in hits2?__ 
There are 140 regions in hist1 and 322 regions in hist2
```{r,eval=TRUE}
dim(as.data.frame(hits1@ranges))[1]
dim(as.data.frame(hits2@ranges))[1]
```
<br>

__Q2. Are hits1 and hits2 proper subsets of all the tested regions? Check how many of each set overlaps a region in all.__

Identical regions in hits1 and hits2 can be found in all. This indicates that both hits1 and hits2 are subsets of all.

Code to identify overlaps
```{r,eval=TRUE}
(findOverlaps(all, hits1, type = "any") %>%
   as.data.frame() %>%
   dim())[1]

(findOverlaps(all, hits2, type = "any") %>%
    as.data.frame() %>%
    dim())[1]
```

Code to identify exact identical regions.
```{r,eval=TRUE}
range_to_df = function(x){
  chr = as.vector(x@seqnames)
  dff = as.data.frame(x@ranges) %>%
    mutate(chr = chr) %>%
    select(chr, start, end, width) %>%
    mutate(chr_start_end = paste(chr, start, end, sep = "_"))
  return(dff)
}

all_df = range_to_df(all)
hits1_df = range_to_df(hits1)
hits2_df = range_to_df(hits2)

all_hits1_merged = inner_join(all_df, select(hits1_df, chr_start_end), 
                              by = "chr_start_end")
all_hits2_merged = inner_join(all_df, select(hits2_df, chr_start_end), 
                              by = "chr_start_end")

dim(all_hits1_merged)[1]
dim(all_hits2_merged)[1]
```

<br>

The next few questions explore the overlap of genomic regions in hits1 and hits2.

__Q3. How many regions overlap? How many regions are exactly identical?__

6 regions overlap in hits1 and hits2 and similarly, 6 regions are exactly identical in hits1 and hits2.
```{r,eval=TRUE}
hits1_hits2_merged = inner_join(hits1_df, hits2_df, by = "chr_start_end")
dim(hits1_hits2_merged)[1]

suppressWarnings({
  (findOverlaps(hits1, hits2, type = "any") %>%
     as.data.frame() %>%
     dim())[1]
})

```
<br>

__Q4. Generate a set of random genomic regions of the same size as hits1. Match these to the mean and sd of the genomic length of the hits1 regions.__ 
<br>
- Do the random genomic regions overlap hits2 more or less than hits1 does? <br>
- How much do the random genomic regions overlap all tested regions? <br>
- Repeatedly generate genomic regions to compute a z-score for the overlap of hits1 with hits2 <br>
- Use the set of overlaps with random regions to test the null hypothesis that hits2 overlaps hits1 more than expected compared to totally random regions <br> 
- What is the smallest p-value you could have gotten? <br>
- How do the results change with number of resamples? Random seed? 
<br>

```{r,eval=TRUE}
genomic_lengths = width(hits1)
mean_length = mean(genomic_lengths)
sd_length = sd(genomic_lengths)

set.seed(492357816)
random_gr = createRandomRegions(nregions = dim(hits1_df)[1], 
                                length.mean = mean_length, 
                                length.sd = sd_length)

numOverlaps(hits1, random_gr)
numOverlaps(hits2, random_gr)

```

Since the random region generated does not overlap with either regions in hits1 or hits2, it is hard to say whether the random regions generated using this method will overlap with hits1 or hits2 more. (Not with further testing)

```{r,eval=TRUE}
# Run perm test using the permTest function
set.seed(492357816)
perm_hits1_result = suppressWarnings({
  permTest(A = hits1, B = hits2, ntimes = 1000, 
           randomize.function = randomizeRegions, 
           evaluate.function = numOverlaps)
})
plot(perm_hits1_result)
```

<br>

```{r,eval=TRUE}
set.seed(492357816)
perm_hits2_result = suppressWarnings({
  permTest(A = hits2, B = hits1, ntimes = 1000, 
           randomize.function = randomizeRegions, 
           evaluate.function = numOverlaps)
})
plot(perm_hits2_result)
```


```{r,eval=TRUE}
# Run perm test using the self-defined function
set.seed(492357816)
hits1_overlap_list = vector()
hits2_overlap_list = vector()
for(i in 1:1000){
  random_gr = suppressWarnings({randomizeRegions(hits1, genome = "hg19")})
  hits1_overlap = numOverlaps(hits1, random_gr)
  hits2_overlap = numOverlaps(hits2, random_gr)
  hits1_overlap_list = c(hits1_overlap_list, hits1_overlap)
  hits2_overlap_list = c(hits2_overlap_list, hits2_overlap)
}

self_perm_result = t.test(hits1_overlap_list, hits2_overlap_list)

perm_hits1_result
perm_hits2_result
self_perm_result
```

Based on the results of both the perTest function and the self-defined permutation test, the random simulated genomic region seems to overlap with regions in hits2 more than it does with hits1. Based on the self-implemented permutation test, the z score is approximately -4.5. This indicates that on average, a randomly generated region overlaps with  regions in hits1 ~4.5 times less than it overlaps with regions in hits2.

<br>

```{r,eval=TRUE}
# t.test(width(hits1), width(hits2))
mean(width(hits1)) * dim(hits1_df)[1]
mean(width(hits2)) * dim(hits2_df)[1]
```

The difference can be explained by the comparison of the product of average region width and region count. hits1 regions are on average wider but there are significantly more regions in hits2, making hits2 cover a broader genomic region in general and therefore landing more hits by chance.

<br>

The smallest p-value that we could have gotten is zero, given that the number of resamples is ~infinitely large.

```{r,eval=TRUE}
run_perm_test = function(random_seed = 492357816){
  set.seed(random_seed)
  hits1_overlap_list = vector()
  hits2_overlap_list = vector()
  for(i in 1:1000){
    random_gr = suppressWarnings({randomizeRegions(hits1, genome = "hg19")})
    hits1_overlap = numOverlaps(hits1, random_gr)
    hits2_overlap = numOverlaps(hits2, random_gr)
    hits1_overlap_list = c(hits1_overlap_list, hits1_overlap)
    hits2_overlap_list = c(hits2_overlap_list, hits2_overlap)
  }
  self_perm_result = t.test(hits1_overlap_list, hits2_overlap_list)
  print(self_perm_result)
}

run_perm_test(43479)
run_perm_test(89669)
run_perm_test(91617)
```

Here three different random seeds are used to run the permutation test. Although the effect size is consistantly negative, it seems to fluctuate between -1.5 to -3.6. This is to say that, given large enough numbers of resamples, random seeds will not change the overlap direction of the test. However, as the number of resamples get larger, the test statistics will become more stable (p-value more significant). On the contrary, when the number of resamples get smaller, the test statistics will be more volatile. 

<br>

__Q5. Repeat Q4 switching the roles of hits1 and hits2. Are conclusions similar?__
```{r,eval=TRUE}
set.seed(492357816)
hits1_overlap_list = vector()
hits2_overlap_list = vector()
for(i in 1:1000){
  random_gr = suppressWarnings({randomizeRegions(hits2, genome = "hg19")})
  hits1_overlap = numOverlaps(hits1, random_gr)
  hits2_overlap = numOverlaps(hits2, random_gr)
  hits1_overlap_list = c(hits1_overlap_list, hits1_overlap)
  hits2_overlap_list = c(hits2_overlap_list, hits2_overlap)
}

self_perm_result = t.test(hits1_overlap_list, hits2_overlap_list)
self_perm_result
```

The conclusions are relatively similar. This suggests that the chances of a randomly generated region overlapping with two genomic regions of interest is largely dependently on the overall size of the regions of interests.


<br>


__Q6. Create a random bootstrap sample of regions from all tested regions.__ <br>
- Do these random regions overlap hits2 more or less than hits1 does? <br>
- How does this test differ from the one in Q4? Look at the z-score and p-value. 

These random regions overlap hits2 more than hits1 does. reasons explained in the previous question.
The p values are comparible but the z score is smaller in this test than what was implemented in the previous tests. 

<br>

```{r,eval=TRUE}
set.seed(492357816)
btsp_hits1_result = permTest(A = hits1, B = hits2, ntimes = 1000, 
                             randomize.function = resampleRegions, 
                             universe = all, 
                             evaluate.function = numOverlaps)
btsp_hits1_result
plot(btsp_hits1_result)
```
<br>


__Q7. Repeat Q6 switching the role of hits1 and hits2. Are conclusions similar?__

The conclusion is very similar.

```{r,eval=TRUE}
set.seed(492357816)
btsp_hits2_result = permTest(A = hits2, B = hits1, ntimes = 1000, 
                             randomize.function = resampleRegions, 
                             universe = all, 
                             evaluate.function = numOverlaps)
btsp_hits2_result
plot(btsp_hits2_result)
```
<br>

__Q8. Which null distribution would you use in your own research and why?__
<br>

If my research primarily concerns regions of interests in hits1, then I would use the NULL distribution generated based on hits1. If I primarily focus on hits2, then I would use the NULL distro simulated using hits2. This way, I can get the most accurate depiction of the distribution of my own regions of interests.

<br>
The next few questions involve downloading genomics data. You can choose sets of regions, e.g, gene annotation, ChIPseq, RNAseq, ATACseq, GWAS SNPs


<br>

__Q9. Using data you download, can you infer what function was tested in the assay that discovered hits1 and hits2? Choose data sets that will be informative about candidate functions. Compute overlaps or mean values of the downloaded data for the union of hits1 and hits2__

data DOI: doi:10.17989/ENCSR000AHO

Human gene annotation file is downloaded from: https://www.gencodegenes.org/human/

```{r,eval=TRUE}
# Clean the dataset
# gtf = fread("gencode.v44.annotation.gtf")
# save(gtf, file = "gtf.rda")
load("gtf.rda")
gtf = as.data.frame(gtf)

names(gtf) = c("chr", "db", "gene", "start", "end", "v6", "strand", "v8", "anno")
gtf = gtf %>%
  select(chr, db, gene, start, end, anno) %>%
  filter(gene == "gene") %>%
  filter(db == "ENSEMBL")

gene_name = strsplit(gtf$anno, split = "gene_name") %>%
  lapply(function(x){x[2]}) %>%
  unlist() %>%
  strsplit(split = ";") %>%
  lapply(function(x){x[1]}) %>%
  unlist() %>% 
  gsub(pattern = "\"", replacement = "") %>%
  gsub(pattern = " ", replacement = "")

gtf = gtf %>%
  mutate(gene_name = gene_name) %>%
  select(chr, db, gene, start, end, gene_name, anno) %>%
  mutate(pos = paste(chr, start, end, sep = "_"))

anno = GRanges(
  seqnames = Rle(gtf$chr),
  ranges = IRanges(start = gtf$start, end = gtf$end),
  strand = "*",
  metadata = tibble(
    gene_name = gtf$gene_name,
    pos = gtf$pos
  )
)

anno_df = as.data.frame(anno)

encode = toGRanges(read.table("ENCFF360QRN.bed",sep="\t"))
overlapRegions(hits1, encode)
overlapRegions(hits2, encode)

overlap_hits_encode = overlapRegions(mergeRegions(hits1, hits2), encode)

annotated = overlapRegions(overlap_hits_encode, anno)
annotated = as.data.frame(annotated) %>%
  mutate(metadata.pos = paste(chr, startB, endB, sep = "_")) %>%
  left_join(anno_df, by = "metadata.pos") %>%
  select(chr, startA, endA, metadata.gene_name)

annotated$metadata.gene_name
```

<br>
__Guess what type of genomic element these hits are (i.e., what assay was performed))__

Based on the fact that the study is a ChIP seq dataset, the overlapped regions are most likely Transcription factor genes. 

<br>
<br>

__BONUS Q10. Do you think hits1 and hits2 function in the same cell type?__ <br>
- Build on your analysis in Q9 by separately testing overlaps with hits1 and hits2. Choose datasets that are from several different cell types 
<br>

```{r,eval=TRUE}
overlapRegions(hits1, encode)
overlapRegions(hits2, encode)

overlapRegions(hits1, anno) %>%
  as.data.frame() %>%
  mutate(metadata.pos = paste(chr, startB, endB, sep = "_")) %>%
  left_join(anno_df, by = "metadata.pos") %>%
  select(chr, startA, endA, metadata.gene_name)

overlapRegions(hits2, anno) %>%
  as.data.frame() %>%
  mutate(metadata.pos = paste(chr, startB, endB, sep = "_")) %>%
  left_join(anno_df, by = "metadata.pos") %>%
  select(chr, startA, endA, metadata.gene_name)

```


In my answer in Q9, I annotated both hits1 and hits2 with genes in the regions. It seems that the number of genes overlapping is not the same. In addition, the overlapped regions seem to be different across the two hits. I therefore suspect that hits1 and hits2 are derived from two different cell types.


Preparing for gene ontology enrichment analysis.
```{r,eval=TRUE}
hits1_genes = overlapRegions(hits1, anno) %>%
  as.data.frame() %>%
  mutate(metadata.pos = paste(chr, startB, endB, sep = "_")) %>%
  left_join(anno_df, by = "metadata.pos") %>%
  select(chr, startA, endA, metadata.gene_name)

hits2_genes = overlapRegions(hits2, anno) %>%
  as.data.frame() %>%
  mutate(metadata.pos = paste(chr, startB, endB, sep = "_")) %>%
  left_join(anno_df, by = "metadata.pos") %>%
  select(chr, startA, endA, metadata.gene_name)

hits1_genes = hits1_genes$metadata.gene_name
hits2_genes = hits2_genes$metadata.gene_name

# paste(unique(hits1_genes), collapse = "\n") %>% cat()
# paste(unique(hits2_genes), collapse = "\n") %>% cat()
```

Gene ontology enrichment analysis was performed on: https://geneontology.org/

The enrichment results for hits1_genes are not significant, however, the results for hits2_genes are significnat, with the top terms being 'neural nucleus development', 'neuron migration', and 'axon guidance'. This suggests that the cell type from hits2 might have something to do with neurons.

<br>

__BONUS Q11: Try matching the random regions more closely to regions in hits1__ <br>
- On what variables will you match them? e.g., spacing, chromosome, GC-content, distance to nearest gene, masking <br>
- How does matching affect the z-score and p-value? <br>

If I were to match the random regions more closely with regions in hits1, I would try to match their spacing, GC-content, distance to nearest gene as well as genomic region (exonic, intronic, 3/5' UTR etc.). Performing closer matching is predicted to increase the z score and making the p-value more significant. This is because closer matching reduces the degrees of freedom of selection of regions in the genome. Therefore making the selected random regions more likely to be similar to regions in hits1, which overlaps more with hits2. 



















